### PR TITLE
New: startAt file option support for media and document viewers

### PR DESCRIPTION
**Pull request by DanDeMicco**

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @DanDeMicco has signed the CLA. Thanks for the pull request!

File: README.md [hunk 1]
```diff
@@ -137,6 +137,7 @@ preview.show(fileId, accessToken, {
 | File Option | Description |
 | --- | --- |
 | fileVersionId | File version ID to preview. This must be a valid non-current file version ID. Use [Get Versions](https://developer.box.com/reference#view-versions-of-a-file) to fetch a list of file versions |
+| start | The unit and value of where to start the preview at. Current supported units are 'seconds' for media and 'pages' for documents. |
```
Comment (pramodsum) [inline README.md]: Maybe we can call this option 'startAt' or something? Start seems a bit confusing

File: functional-tests/fileOptions_test.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
+const { SELECTOR_BOX_PREVIEW_LOADED, SELECTOR_MEDIA_TIMESTAMP, SELECTOR_DOC_CURRENT_PAGE } = require('./constants');
```
Comment (pramodsum) [inline functional-tests/fileOptions_test.js]: LOVE that there's functional tests already for this!

File: functional-tests/fileOptions_test.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+const MP4_START = '0:10';
+
+Feature('File Options', { retries: CI ? 3 : 0 });
+
+Before((I) => {
+    I.amOnPage('/functional-tests/file-options.html');
+    I.waitForElement(SELECTOR_BOX_PREVIEW_LOADED);
+});
+
+Scenario('Check preview starts at correct spot for all file types @ci', (I) => {
```
Comment (pramodsum) [inline functional-tests/fileOptions_test.js]: split these up into multiple scenarios?

File: src/lib/Fullscreen.js [hunk 4]
```diff
@@ -28,7 +28,7 @@ class Fullscreen extends EventEmitter {
     }
 
     /**
-     * Binds DOM listeners for Fullscreen.
+     * Unbinds DOM listeners for Fullscreen.
```
Comment (pramodsum) [inline src/lib/Fullscreen.js]: good catch

File: src/lib/viewers/BaseViewer.js [hunk 5]
```diff
@@ -108,6 +109,10 @@ class BaseViewer extends EventEmitter {
         this.options = options;
         this.cache = options.cache;
         this.previewUI = options.ui;
+        const fileOptions = options.fileOptions || {};
```
Comment (pramodsum) [inline src/lib/viewers/BaseViewer.js]: Move these into a section below w/ a new line above it

File: src/lib/viewers/doc/DocBaseViewer.js [hunk 6]
```diff
@@ -153,6 +155,33 @@ class DocBaseViewer extends BaseViewer {
...
+     * Converts a value and unit to page number
+     *
+     * @return {number|undefined} a page number > 0
+     */
+    getStartPage() {
+        let convertedValue;
+
+        const { unit, value } = this.fileStartObj || {};
+
+        if (value && unit) {
```
Comment (pramodsum) [inline src/lib/viewers/doc/DocBaseViewer.js]:
```
let convertedValue = null;
...
if (!value || !unit) {
    return convertedValue
}
...
```

File: src/lib/viewers/doc/DocBaseViewer.js [hunk 7]
```diff
@@ -70,6 +71,7 @@ class DocBaseViewer extends BaseViewer {
         this.pinchToZoomStartHandler = this.pinchToZoomStartHandler.bind(this);
         this.pinchToZoomChangeHandler = this.pinchToZoomChangeHandler.bind(this);
         this.pinchToZoomEndHandler = this.pinchToZoomEndHandler.bind(this);
+        this.startPageNum = this.getStartPage();
```
Comment (pramodsum) [inline src/lib/viewers/doc/DocBaseViewer.js]: Move this above the function bindings

File: src/lib/viewers/media/MediaBaseViewer.js [hunk 8]
```diff
@@ -60,6 +63,58 @@ class MediaBaseViewer extends BaseViewer {
...
+     * @param {string} unit - the unit e.g. seconds
+     * @return {number} a time in seconds
+     */
+    getStartTimeInSeconds() {
+        let convertedValue = 0;
+
+        const value = getProp(this.fileStartObj, 'value');
+        const unit = getProp(this.fileStartObj, 'unit');
+
+        if (value && unit) {
```
Comment (pramodsum) [inline src/lib/viewers/media/MediaBaseViewer.js]:
```
if (!value || !unit) {
    return;
}
...
```

File: src/lib/viewers/doc/DocBaseViewer.js [hunk 9]
```diff
@@ -153,6 +155,33 @@ class DocBaseViewer extends BaseViewer {
         super.destroy();
     }
 
+    /**
+     * Converts a value and unit to page number
+     *
+     * @return {number|undefined} a page number > 0
+     */
+    getStartPage() {
```
Comment (jeremypress) [inline src/lib/viewers/doc/DocBaseViewer.js]: I recommend calling this in pagesInitHandler where we already handle logic like this. We can also do validation at the same time, and fall back to a cached value that we may have stored.

File: src/lib/viewers/doc/DocBaseViewer.js [hunk 10]
```diff
@@ -153,6 +155,33 @@ class DocBaseViewer extends BaseViewer {
...
+        let convertedValue;
+
+        const { unit, value } = this.fileStartObj || {};
+
+        if (value && unit) {
+            switch (unit) {
+                case PAGES_UNIT_NAME:
+                    convertedValue = parseInt(value, 10);
+                    if (Number.isNaN(convertedValue) || convertedValue < 1) {
```
Comment (jeremypress) [inline src/lib/viewers/doc/DocBaseViewer.js]: NAN is already falsy

File: src/lib/viewers/media/MediaBaseViewer.js [hunk 11]
```diff
@@ -60,6 +63,58 @@ class MediaBaseViewer extends BaseViewer {
...
+
+        this.startTimeInSeconds = this.getStartTimeInSeconds();
+    }
+
+    /**
+     * Sets the start time for a media element
+     *
+     * @return {void}
+     */
+    loadedMetadataHandler() {
```
Comment (jeremypress) [inline src/lib/viewers/media/MediaBaseViewer.js]: Did we not end up needing to switch to `loadeddata`?

File: src/lib/Preview.js [hunk 12]
```diff
@@ -824,6 +824,9 @@ class Preview extends EventEmitter {
         // (access stats will not be incremented), but content access is still logged server-side for audit purposes
         this.options.disableEventLog = !!options.disableEventLog;
 
+        // Options that are applicable to certain file id's
```
Comment (JustinHoldstock) [inline src/lib/Preview.js]: you can remove the apostrophe in "id's"

File: src/lib/viewers/doc/DocBaseViewer.js [hunk 13]
```diff
@@ -483,6 +514,19 @@ class DocBaseViewer extends BaseViewer {
...
 
+    /**
+     * Sets the start page the preview should start at
+     *
+     * @param {number} startPageNum - the start page number to start the preview at
+     * @param {number} pagesCount - the number of pages in the document
+     * @return {void}
+     */
+    setStartPage(startPageNum, pagesCount) {
+        if (startPageNum && startPageNum > 0 && startPageNum <= pagesCount) {
```
Comment (JustinHoldstock) [inline src/lib/viewers/doc/DocBaseViewer.js]: Would it be safer to `parseInt(startPageNum)` and check to make sure it isn't `NaN`? `true` can cause this logic to run, as `startPageNum > 0 && startPageNum <= pagesCount` equates to `true`, where `startPageNum = true`. Also, as a bonus, we'll have floats and strings parsed to integers and avoid passing strange data to `this.cachePage()`

File: README.md [hunk 14]
```diff
@@ -137,6 +137,7 @@ preview.show(fileId, accessToken, {
 | File Option | Description |
 | --- | --- |
 | fileVersionId | File version ID to preview. This must be a valid non-current file version ID. Use [Get Versions](https://developer.box.com/reference#view-versions-of-a-file) to fetch a list of file versions |
+| startAt | The unit and value of where to start the preview at. Current supported units are 'seconds' for media and 'pages' for documents. |
```
Comment (tonyjin) [inline README.md]: `Object with unit and value properties indicating where to start the preview at. ...`

File: src/lib/viewers/media/MediaBaseViewer.js [hunk 15]
```diff
@@ -60,6 +63,49 @@ class MediaBaseViewer extends BaseViewer {
...
+
+        const value = getProp(this.fileStartObj, 'value');
+        const unit = getProp(this.fileStartObj, 'unit');
+
+        if (!value || !unit) {
+            return INITIAL_TIME_IN_SECONDS;
+        }
+
+        if (unit === SECONDS_UNIT_NAME) {
+            convertedValue = parseInt(value, 10);
```
Comment (JustinHoldstock) [inline src/lib/viewers/media/MediaBaseViewer.js]: Does this mean that we can't start at fractions of a second? Or does Shaka player not support that?

File: src/lib/viewers/BaseViewer.js [hunk 16]
```diff
@@ -108,6 +109,12 @@ class BaseViewer extends EventEmitter {
         this.options = options;
         this.cache = options.cache;
         this.previewUI = options.ui;
+
+        const fileOptions = options.fileOptions || {};
```
Comment (tonyjin) [inline src/lib/viewers/BaseViewer.js]: const { fileOptions } = options;

You already default fileOptions to {} in Preview.js

File: src/lib/viewers/media/MediaBaseViewer.js [hunk 17]
```diff
@@ -60,6 +63,49 @@ class MediaBaseViewer extends BaseViewer {
...
+        return convertedValue;
+    }
+
+    /**
+     * Takes the start parameter, validates, converts, and applies to the media element
+     *
+     * @return {void}
+     */
+    setStartTime() {
+        if (this.mediaEl && this.startTimeInSeconds > 0 && this.startTimeInSeconds < this.mediaEl.duration) {
```
Comment (JustinHoldstock) [inline src/lib/viewers/media/MediaBaseViewer.js]: I would prefer that this function acts as a traditional mutator, and accepts a parameter. Wherever you call it and require `this.startTimeInSeconds`, just pass it in. It'll be more flexible that way, as it seems strange to have to set the value of `startTimeInSeconds`, and then additionally call `setStartTime()`.

Another note, instead of calling this `setStartTime()`, would it be better to just do the validation inside of `setMediaTime()` and call it, with `this.startTimeInSeconds`?

File: src/lib/viewers/BaseViewer.js [hunk 18]
```diff
@@ -108,6 +109,12 @@ class BaseViewer extends EventEmitter {
...
+        const fileId = getProp(options, 'file.id');
+        const optionsForFileId = fileOptions[fileId] || {};
+        this.fileStartObj = optionsForFileId[FILE_OPTION_START];
```
Comment (tonyjin) [inline src/lib/viewers/BaseViewer.js]: I think you can just do `this.startAt = getProp(options, `fileOptions.${options.file.id}`, {});`

I like keeping `startAt` since that's what we called the Preview-level option.

File: src/lib/viewers/doc/DocBaseViewer.js (hunk 9 above)
Comment (tonyjin) [reply src/lib/viewers/doc/DocBaseViewer.js]: +1 - see in pagesinitHandler():

```
// Set current page to previously opened page or first page
        this.setPage(this.getCachedPage());
```

Like Jeremy mentioned, `startAt` should take precedence over whatever value is stored in local storage, and if there's no `startAt`, fall back to page in local storage.

File: src/lib/viewers/doc/DocBaseViewer.js [hunk 19]
```diff
@@ -153,6 +156,33 @@ class DocBaseViewer extends BaseViewer {
...
 
+    /**
+     * Converts a value and unit to page number
+     *
+     * @return {number|undefined} a page number > 0
+     */
+    getStartPage() {
+        let convertedValue;
+
+        const { unit, value } = this.fileStartObj || {};
```
Comment (tonyjin) [inline src/lib/viewers/doc/DocBaseViewer.js]: If you do `this.startAt = getProp(options,fileOptions.${options.file.id}, {});` above, you don't need to default again to {} here.

File: src/lib/viewers/media/MediaBaseViewer.js (hunk 15 above)
Comment (DanDeMicco) [reply src/lib/viewers/media/MediaBaseViewer.js]: Good point. It definitely supports floating point! changing to parseFloat

File: src/lib/viewers/doc/DocBaseViewer.js [hunk 20]
```diff
@@ -153,6 +156,33 @@ class DocBaseViewer extends BaseViewer {
...
+            convertedValue = parseInt(value, 10);
+            if (!convertedValue || convertedValue < 1) {
+                // Negative values aren't allowed, fall back to default behavior
+                return undefined;
+            }
+        } else {
+            console.error('Invalid unit for start:', unit); // eslint-disable-line no-console
+        }
+
+        return convertedValue;
```
Comment (tonyjin) [inline src/lib/viewers/doc/DocBaseViewer.js]: I don't quite remember, but if you provide a page number past the last page to our setPage function, that gracefully handles the edge case right?

File: src/lib/viewers/doc/DocBaseViewer.js (hunk 13 above)
Comment (tonyjin) [reply src/lib/viewers/doc/DocBaseViewer.js]: Also, I don't think you need to have any of this logic at all. `setPage()` internally manages valid pages. If you have a value for page, it's safe to blindly pass that to `setPage()`

File: src/lib/viewers/doc/DocBaseViewer.js [hunk 21]
```diff
@@ -866,6 +910,13 @@ class DocBaseViewer extends BaseViewer {
 
         this.loadUI();
 
+        const { pagesCount, currentScale } = this.pdfViewer;
+
+        // Set current page to the value passed for startAt
+        if (this.startPageNum) {
+            this.setStartPage(this.startPageNum, pagesCount);
+        }
```
Comment (tonyjin) [inline src/lib/viewers/doc/DocBaseViewer.js]: Move this logic to `pagesinitHandler()`, should be something like:

```
// Start page at user-defined page, the last cached page, or 1
const startPage = this.getStartPage() || this.getCachedPage();
this.setPage(startPage);
```

File: src/lib/viewers/media/MediaBaseViewer.js [hunk 22]
```diff
@@ -60,6 +63,49 @@ class MediaBaseViewer extends BaseViewer {
         this.loadTimeout = 100000;
         this.oldVolume = DEFAULT_VOLUME;
         this.pauseListener = null;
+
+        this.startTimeInSeconds = this.getStartTimeInSeconds();
```
Comment (tonyjin) [inline src/lib/viewers/media/MediaBaseViewer.js]: Ah I see, you need to set it as a property in MediaBase. In that case, it may be worth keeping consistency in the doc viewer.

Then the doc logic becomes:
`const startPage = this.startPageNum || this.getCachedPage();`

File: src/lib/viewers/BaseViewer.js (hunk 18 above)
Comment (tonyjin) [reply src/lib/viewers/BaseViewer.js]: Whatever you decide to name this, add it as a property defined above the constructor. We decided a while back we liked keeping track of all properties and what they're used for at the top of the Class.

File: src/lib/viewers/media/MediaBaseViewer.js (hunk 17 above)
Comment (tonyjin) [reply src/lib/viewers/media/MediaBaseViewer.js]: +1, that way, you mirror what's done in the doc viewer (validation in `setPage()`, pass in `this.startPageNum`)

File: src/lib/viewers/doc/DocBaseViewer.js (hunk 19 above)
Comment (DanDeMicco) [reply src/lib/viewers/doc/DocBaseViewer.js]: True, was just being doubly sure

File: src/lib/viewers/BaseViewer.js (hunk 18 above)
Comment (DanDeMicco) [reply src/lib/viewers/BaseViewer.js]: My intention was to get all the file options and store them, but separating into individual properties on the viewer also makes sense and makes things more readable

File: src/lib/viewers/media/MediaBaseViewer.js (hunk 17 above)
Comment (DanDeMicco) [reply src/lib/viewers/media/MediaBaseViewer.js]: Yeah, I guess this method can be totally removed and the logic combined in ```setMediaTime```

File: src/lib/Preview.js (hunk 12 above)
Comment (DanDeMicco) [reply src/lib/Preview.js]: 👍

File: src/lib/viewers/doc/DocBaseViewer.js (hunk 9 above)
Comment (DanDeMicco) [reply src/lib/viewers/doc/DocBaseViewer.js]: Moved the call to pagesInitHandler

File: src/lib/viewers/doc/DocBaseViewer.js [hunk 23]
```diff
@@ -234,6 +264,7 @@ class DocBaseViewer extends BaseViewer {
     load() {
         this.setup();
         super.load();
+
```
Comment (jeremypress) [inline src/lib/viewers/doc/DocBaseViewer.js]: extra line?

File: src/lib/viewers/doc/DocBaseViewer.js [hunk 24]
```diff
@@ -862,8 +894,11 @@ class DocBaseViewer extends BaseViewer {
 
         this.loadUI();
 
-        // Set current page to previously opened page or first page
-        this.setPage(this.getCachedPage());
+        const { pagesCount, currentScale } = this.pdfViewer;
+
+        // Set page to the user-defined page, previously opened page, or first page
+        const startPage = this.startPageNum || this.getCachedPage();
```
Comment (jeremypress) [inline src/lib/viewers/doc/DocBaseViewer.js]: You're not using this.startPageNum anywhere else, why not just set it here locally instead of a class variable?

`const startPage = this.getStartPage() || this.getCachedPage();`

Edit: Just saw Tony's comment, that makes sense. Would it make sense to standardize on the name across viewers?

File: src/lib/viewers/media/DashViewer.js [hunk 25]
```diff
@@ -150,6 +150,7 @@ class DashViewer extends VideoBaseViewer {
         this.adapting = true;
         this.player = new shaka.Player(this.mediaEl);
         this.player.addEventListener('adaptation', this.adaptationHandler);
+
```
Comment (jeremypress) [inline src/lib/viewers/media/DashViewer.js]: extra line?

File: functional-tests/header_test.js [hunk 26]
```diff
@@ -1,13 +1,15 @@
 const { SELECTOR_BOX_PREVIEW_BTN_DOWNLOAD, SELECTOR_DOWNLOAD_IFRAME } = require('./constants');
 const { expect } = require('chai');
 
-Feature('Header', { retries: 3 });
```
Comment (pramodsum) [inline functional-tests/header_test.js]: I would move these into a separate PR

File: functional-tests/helpers.js [hunk 27]
```diff
@@ -0,0 +1,70 @@
+const { SELECTOR_BOX_PREVIEW, CLASS_BOX_PREVIEW_LOADING_WRAPPER } = require('./constants');
```
Comment (pramodsum) [inline functional-tests/helpers.js]: If this is unrelated to the startAt commits, I'd move this into a separate PR

File: functional-tests/sanity_test.js [hunk 28]
```diff
@@ -5,7 +5,7 @@ Before((I) => {
     I.amOnPage('/functional-tests/index.html');
 });
 
-Scenario('Sanity test @ci @mobile', (I) => {
```
Comment (pramodsum) [inline functional-tests/sanity_test.js]: separate PR?

File: src/lib/polyfill.js [hunk 29]
```diff
@@ -210,4 +210,11 @@ if (!String.prototype.endsWith) {
         }
     });
 }
+
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isNaN
+Number.isNaN =
```
Comment (pramodsum) [inline src/lib/polyfill.js]: 👍

File: src/lib/viewers/doc/DocBaseViewer.js [hunk 30]
```diff
@@ -154,6 +157,33 @@ class DocBaseViewer extends BaseViewer {
...
+
+        const { unit, value } = this.startAt;
+
+        if (!value || !unit) {
+            return convertedValue;
+        }
+
+        if (unit === PAGES_UNIT_NAME) {
+            convertedValue = parseInt(value, 10);
+            if (!convertedValue || convertedValue < 1) {
```
Comment (pramodsum) [inline src/lib/viewers/doc/DocBaseViewer.js]: new line before if statement

File: src/lib/viewers/media/MediaBaseViewer.js [hunk 31]
```diff
@@ -354,7 +390,9 @@ class MediaBaseViewer extends BaseViewer {
      * @return {void}
      */
     setMediaTime(time) {
-        this.mediaEl.currentTime = time;
+        if (this.mediaEl && time >= 0 && time <= this.mediaEl.duration) {
```
Comment (jeremypress) [inline src/lib/viewers/media/MediaBaseViewer.js]: Does anything weird happen if you set it to exactly the duration?

File: src/lib/viewers/doc/DocBaseViewer.js [hunk 32]
```diff
@@ -154,6 +157,33 @@ class DocBaseViewer extends BaseViewer {
...
+                // Negative values aren't allowed, fall back to default behavior
+                return undefined;
+            }
+        } else {
+            console.error('Invalid unit for start:', unit); // eslint-disable-line no-console
```
Comment (pramodsum) [inline src/lib/viewers/doc/DocBaseViewer.js]: Should we be surfacing this error in some other way other than as a console.error?

**Comment by jeremypress** [comment]
In the future please break stuff like the functional tests into their own PR since they aren't related.

File: src/lib/viewers/media/MediaBaseViewer.js [hunk 33]
```diff
@@ -60,6 +63,38 @@ class MediaBaseViewer extends BaseViewer {
...
+        const value = getProp(this.startAt, 'value');
+        const unit = getProp(this.startAt, 'unit');
+
+        if (!value || !unit) {
+            return INITIAL_TIME_IN_SECONDS;
+        }
+
+        if (unit === SECONDS_UNIT_NAME) {
+            convertedValue = parseFloat(value, 10);
+            if (convertedValue < 0) {
```
Comment (pramodsum) [inline src/lib/viewers/media/MediaBaseViewer.js]: new line before if statement

File: src/lib/viewers/media/MediaBaseViewer.js [hunk 34]
```diff
@@ -60,6 +63,38 @@ class MediaBaseViewer extends BaseViewer {
...
+                // Negative values aren't allowed, start from beginning
+                return INITIAL_TIME_IN_SECONDS;
+            }
+        } else {
+            console.error('Invalid unit for start:', unit); // eslint-disable-line no-console
```
Comment (pramodsum) [inline src/lib/viewers/media/MediaBaseViewer.js]: should we be surfacing this error through a console.error? Or somethign else

File: functional-tests/fileOptions_test.js [hunk 35]
```diff
@@ -0,0 +1,102 @@
+const {
+    SELECTOR_BOX_PREVIEW_LOADED,
+    SELECTOR_MEDIA_TIMESTAMP,
+    SELECTOR_DOC_CURRENT_PAGE,
+    SELECTOR_BOX_PREIVIEW_LOGO,
```
Comment (pramodsum) [inline functional-tests/fileOptions_test.js]: spelling
